javascript/extractor/src/com/semmle/jcorn/CustomParser.java

```diff
@@ -546,12 +546,21 @@ protected XMLAttributeSelector decoratorToAttributeSelector(Decorator d) {
 
 	@Override
 	protected Token readToken(int code) {
-		// skip XML processing instructions (which are allowed in E4X, but not in JSX)
+		// skip XML processing instructions (which are allowed in E4X, but not in JSX);
+		// there is a lexical ambiguity between an XML processing instruction starting a
+		// chunk of E4X content and a Flow type annotation (both can start with `<?`)
+		// hence if we can't find the closing `?>` of a putative XML processing instruction
+		// we backtrack and try lexing as something else
```

Comment:
Another idea for a dirty workaround: lookahead for the next `>`, and if that's preceded by `?` assume E4X otherwise Flow. A Flow type annotation will almost never contain `?>` and an XML processing instruction can't (?) contain a lone `>`, as it would need to be escaped as `&gt;`.

Reply:
Another idea: can we commit to a parser mode for the rest of the compilation unit once we have observed the first evidence of a specific language variant? Then the lookahead problem will only occur if `<?` comes before all other evidence.

Reply:
I've implemented the first proposal, restricting it to `<?xml`.